src/full/Agda/TypeChecking/Records.hs

```diff
@@ -513,7 +513,17 @@ isEtaRecordConstructor c = isRecordConstructor c >>= \case
 -- | Turn off eta for unguarded recursive records.
 --   Projections do not preserve guardedness.
 unguardedRecord :: QName -> PatternOrCopattern -> TCM ()
-unguardedRecord q pat = modifyRecEta q \ eta -> setEtaEquality eta $ NoEta pat
+unguardedRecord q pat =
+  (stSignature . sigDefinitions . lensDefinition q . lensTheDef . lensRecord . lensRecEta)
+    `modifyTCLensM` \case
+      YesEtaPragma            -> return YesEtaPragma -- cannot overwrite
+      e@(Specified _ NoEta{}) -> pure e
+      Specified r YesEta      -> setCurrentRange r do
+        warning $ IllicitEtaRecord Inductive q
```

Comment:
Something is weird about our API for warnings.
This call to `warning` does add deadcode highlighting, but no warning text is displayed.

Reply:
After many hours of unsuccessful search in our warning handling, out of despair I rewrote this code to not use `HashMap.alterF`, but a separate lookup and update.  And lo and behold, the warning text appears.
So it seems that `alterF` partially slanders the effects of `warning`, not the highlighting effect, but the storing of the warning.  WFT?!
More fuel on my aversion against hashmaps.
Hiding an imperative algorithm in IO behind a functional API, that is blowing up more often than hoped, it seems.

Reply:
Er, I have to take this back.  Blame goes to `stSignature` which will lose changes to `TCState` performed by the effectful computation we can pass it.  This seems to be a general issue with van Laarhoven lenses.
So our auxiliary function `modifyTCLensM` is not safe to use: https://github.com/agda/agda/blob/0be54d7a9359cd5898837d5f5f4df49edab99ec8/src/full/Agda/TypeChecking/Monad/Base.hs#L5527-L5530
It is only safe to use with updaters that do not modify the `TCState` as a side effect.  So we would need to take away the `MonadTCState` capability from the _first_ occurrence of `m` (in the update function), but retain it in the _second_ occurrrence (the `modifyTCLensM` function itself).  The Haskell type system does not support this, though.

Reply:
Well that's quite a discovery! So modifying the state through both a lens and an implicit side effect at the same time is not safe. Well, at least now we know what we should not do.

Reply:
To be fair to lenses, the problem is that `modifyTCLensM` is implemented incorrectly here. And I would argue that this foot gun is more about stateful programming rather than lenses. Any time you see `get >>= ... >>= put` you need to ask yourself if you really wanted to discard the state updates in `...`